safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,473 @@
+use anyhow::{bail, ensure, Context, Result};
+
+use std::cmp::min;
+use std::path::{Path, PathBuf};
+use std::time::Duration;
+
+use postgres_ffi::xlog_utils::{XLogFileName, XLogSegNo, XLogSegNoOffsetToRecPtr, PG_TLI};
+use remote_storage::{GenericRemoteStorage, RemoteStorage};
+use tokio::fs::File;
+use tokio::runtime::{Builder, Runtime};
+
+use tokio::select;
+use tokio::sync::watch::{self, Receiver, Sender};
+use tokio::time::sleep;
+use tracing::*;
+
+use utils::{lsn::Lsn, zid::ZTenantTimelineId};
+
+use crate::broker::Election;
+use crate::{broker, SafeKeeperConf};
+
+use once_cell::sync::OnceCell;
+
+static BACKUP_RUNTIME: OnceCell<Runtime> = OnceCell::new();
+const DEFAULT_BACKUP_RUNTIME_SIZE: u32 = 16;
+
+const MIN_WAL_SEGMENT_SIZE: usize = 1 << 20; // 1MB
+const MAX_WAL_SEGMENT_SIZE: usize = 1 << 30; // 1GB
+
+const BACKUP_ELECTION_NAME: &str = "WAL_BACKUP";
+
+const BROKER_CONNECTION_RETRY_DELAY_MS: u64 = 1000;
+
+const UPLOAD_FAILURE_RETRY_MIN_MS: u64 = 10;
+const UPLOAD_FAILURE_RETRY_MAX_MS: u64 = 5000;
+
+#[allow(clippy::too_many_arguments)]
+async fn backup_task(
+    backup_start: Lsn,
+    timeline_id: ZTenantTimelineId,
+    timeline_dir: PathBuf,
+    mut segment_size_set: Receiver<u32>,
+    mut lsn_durable: Receiver<Lsn>,
+    mut shutdown: Receiver<bool>,
+    lsn_backed_up: Sender<Lsn>,
+    election: Election,
+) -> Result<()> {
+    info!("Starting backup task, backup_lsn {}", backup_start);
+
+    segment_size_set
+        .changed()
+        .await
+        .expect("Failed to recieve wal segment size");
+    let wal_seg_size = *segment_size_set.borrow() as usize;
+    ensure!(
+        (MIN_WAL_SEGMENT_SIZE..=MAX_WAL_SEGMENT_SIZE).contains(&wal_seg_size),
+        "Invalid wal seg size provided, should be between 1MiB and 1GiB per postgres"
+    );
+
+    let mut backup_lsn = backup_start;
+
+    loop {
+        let mut leader = None;
+        let mut retry_attempt = 0u64;
+
+        select! {
+            result = broker::get_leader(&election) => {
+                match result {
+                    Ok(l) => { leader = Some(l);},
+                    Err(e) => {
+                        error!("Error during leader election {:?}", e);
+                        sleep(Duration::from_millis(BROKER_CONNECTION_RETRY_DELAY_MS)).await;
+                        continue;
+                    },
+                }
+            }
+            _ = shutdown.changed() => {}
```

Comment:
As discussed, there can be just single top-level select over shutdown chan and the rest. And we can do cleanup and exit right in the arm handling.